tape/src/main/java/com/squareup/tape2/QueueFile.java

```diff
@@ -50,44 +50,58 @@
  * @author Bob Lee (bob@squareup.com)
  */
 public final class QueueFile implements Closeable, Iterable<byte[]> {
+  /** Leading bit set to 1 indicating a versioned header and the version of 1. */
+  private static final int VERSIONED_HEADER = 0x80000001;
+
   /** Initial file size in bytes. */
-  private static final int INITIAL_LENGTH = 4096; // one file system block
+  static final int INITIAL_LENGTH = 4096; // one file system block
 
   /** A block of nothing to write over old data. */
   private static final byte[] ZEROES = new byte[INITIAL_LENGTH];
 
-  /** Length of header in bytes. */
-  static final int HEADER_LENGTH = 16;
-
   /**
    * The underlying file. Uses a ring buffer to store entries. Designed so that a modification
    * isn't committed or visible until we write the header. The header is much smaller than a
    * segment. So long as the underlying file system supports atomic segment writes, changes to the
    * queue are atomic. Storing the file length ensures we can recover from a failed expansion
    * (i.e. if setting the file length succeeds but the process dies before the data can be copied).
-   * <p/>
+   * <p>
+   * This implementation supports two versions of the on-disk format.
    * <pre>
-   *   Format:
-   *     Header              (16 bytes)
-   *     Element Ring Buffer (File Length - 16 bytes)
-   * <p/>
-   *   Header:
-   *     File Length            (4 bytes)
-   *     Element Count          (4 bytes)
-   *     First Element Position (4 bytes, =0 if null)
-   *     Last Element Position  (4 bytes, =0 if null)
-   * <p/>
-   *   Element:
-   *     Length (4 bytes)
-   *     Data   (Length bytes)
-   * </pre>
+   * Format:
+   *   16-32 bytes      Header
+   *   ...              Data
+   *
+   * Header (32 bytes):
+   *   1 bit            Versioned indicator [0 = legacy (see "Legacy Header"), 1 = versioned]
+   *   31 bits          Version, always 1
```

Comment:
31 bits seems excessive. I think 15 is sufficient and we could use a short instead. Though the overhead of adding code to deal with short just for this might not be worth it.